### PR TITLE
DefCoinSelector: prevent NRE when null Random is used

### PR DESCRIPTION
**Pull request by knocte**
Related to issue #1001.

### CODE REVIEW

**Comment by knocte** [comment]
CI failure seems unrelated:

```
[xUnit.net 00:04:57.71]     NBitcoin.Tests.ProtocolTests.CanMaintainChainWithChainBehavior [FAIL]
[xUnit.net 00:04:57.71]       System.InvalidOperationException : Invalid Node state, needed=HandShaked, current= Failed
[xUnit.net 00:04:57.71]       Stack Trace:
[xUnit.net 00:04:57.71]         /Users/runner/work/NBitcoin/NBitcoin/NBitcoin/Protocol/Node.cs(1647,0): at NBitcoin.Protocol.Node.AssertState(NodeState nodeState, CancellationToken cancellationToken)
[xUnit.net 00:04:57.71]         /Users/runner/work/NBitcoin/NBitcoin/NBitcoin/Protocol/Node.cs(1342,0): at NBitcoin.Protocol.Node.GetHeadersFromFork(ChainedBlock currentTip, SynchronizeChainOptions options, CancellationToken cancellationToken)+MoveNext()
[xUnit.net 00:04:57.71]            at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
[xUnit.net 00:04:57.71]            at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
[xUnit.net 00:04:57.71]         /Users/runner/work/NBitcoin/NBitcoin/NBitcoin/Protocol/Node.cs(1444,0): at NBitcoin.Protocol.Node.SynchronizeChain(ChainBase chain, SynchronizeChainOptions options, CancellationToken cancellationToken)
[xUnit.net 00:04:57.71]         /Users/runner/work/NBitcoin/NBitcoin/NBitcoin/Protocol/Node.cs(1309,0): at NBitcoin.Protocol.Node.GetChain(SynchronizeChainOptions options, CancellationToken cancellationToken)
[xUnit.net 00:04:57.71]         /Users/runner/work/NBitcoin/NBitcoin/NBitcoin/Protocol/Node.cs(1334,0): at NBitcoin.Protocol.Node.GetChain(uint256 hashStop, CancellationToken cancellationToken)
[xUnit.net 00:04:57.71]         /Users/runner/work/NBitcoin/NBitcoin/NBitcoin.Tests/ProtocolTests.cs(536,0): at NBitcoin.Tests.ProtocolTests.CanMaintainChainWithChainBehavior()
  Failed NBitcoin.Tests.ProtocolTests.CanMaintainChainWithChainBehavior [1 m 12 s]
  Error Message:
   System.InvalidOperationException : Invalid Node state, needed=HandShaked, current= Failed
  Stack Trace:
     at NBitcoin.Protocol.Node.AssertState(NodeState nodeState, CancellationToken cancellationToken) in /Users/runner/work/NBitcoin/NBitcoin/NBitcoin/Protocol/Node.cs:line 1647
   at NBitcoin.Protocol.Node.GetHeadersFromFork(ChainedBlock currentTip, SynchronizeChainOptions options, CancellationToken cancellationToken)+MoveNext() in /Users/runner/work/NBitcoin/NBitcoin/NBitcoin/Protocol/Node.cs:line 1342
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at NBitcoin.Protocol.Node.SynchronizeChain(ChainBase chain, SynchronizeChainOptions options, CancellationToken cancellationToken) in /Users/runner/work/NBitcoin/NBitcoin/NBitcoin/Protocol/Node.cs:line 1444
   at NBitcoin.Protocol.Node.GetChain(SynchronizeChainOptions options, CancellationToken cancellationToken) in /Users/runner/work/NBitcoin/NBitcoin/NBitcoin/Protocol/Node.cs:line 1309
   at NBitcoin.Protocol.Node.GetChain(uint256 hashStop, CancellationToken cancellationToken) in /Users/runner/work/NBitcoin/NBitcoin/NBitcoin/Protocol/Node.cs:line 1334
   at NBitcoin.Tests.ProtocolTests.CanMaintainChainWithChainBehavior() in /Users/runner/work/NBitcoin/NBitcoin/NBitcoin.Tests/ProtocolTests.cs:line 536
```

**Comment by NicolasDorier** [comment]
In order to not make that a breaking change can you instead, if `int seed == 0`, then do not initialize the Random?

**Comment by NicolasDorier** [comment]
In order to prevent breaking change in the interface, can you remove the new constructor you did. You can already pass `null` to the `DefaultCoinSelection(Random)`

**Comment by knocte** [comment]
> In order to prevent breaking change in the interface, can you remove the new constructor you did. You can already pass null to the DefaultCoinSelection(Random)

Good point, sorry I overlooked that! I'll fix shortly.

**Comment by knocte** [comment]
Updated, but don't merge yet please, I want to check something first.

File: NBitcoin/TransactionBuilder.cs [hunk 1]
```diff
@@ -78,7 +78,7 @@ public DefaultCoinSelector()
 			_Rand = new Random();
 		}
 
-		Random? _Rand;
```
Comment (NicolasDorier) [inline NBitcoin/TransactionBuilder.cs]: should be `Random?`

Comment (knocte) [reply NBitcoin/TransactionBuilder.cs]: but then why do `Random` works? I guess it means you don't have the new null-detection feature of C# enabled?

**Comment by knocte** [comment]
And how can this PR in its current form cause the following test to fail? (see CI log):


```
[xUnit.net 00:08:53.71]     NBitcoin.Tests.transaction_tests.DoNotBuildTooBigTransaction [FAIL]
[xUnit.net 00:08:53.71]       Assert.Equal() Failure
[xUnit.net 00:08:53.71]       Expected: 1 Sat/B
[xUnit.net 00:08:53.71]       Actual:   1.009 Sat/B
[xUnit.net 00:08:53.71]       Stack Trace:
[xUnit.net 00:08:53.72]         transaction_tests.cs(2486,0): at NBitcoin.Tests.transaction_tests.DoNotBuildTooBigTransaction()
  X NBitcoin.Tests.transaction_tests.DoNotBuildTooBigTransaction [104ms]
  Error Message:
   Assert.Equal() Failure
Expected: 1 Sat/B
Actual:   1.009 Sat/B
  Stack Trace:
     at NBitcoin.Tests.transaction_tests.DoNotBuildTooBigTransaction() in D:\a\NBitcoin\NBitcoin\NBitcoin.Tests\transaction_tests.cs:line 2486
```

File: NBitcoin/TransactionBuilder.cs (hunk 1 above)
Comment (knocte) [reply NBitcoin/TransactionBuilder.cs]: > I guess it means you don't have the new null-detection feature of C# enabled?

Confirmed; to opt-in for this we should add `<Nullable>enable</Nullable>` to the .csproj file.

**Comment by NicolasDorier** [comment]
probably not this PR.

**Comment by knocte** [comment]
Thanks for merging, @NicolasDorier can you bump version?

**Comment by NicolasDorier** [comment]
done

**Pull request merged**